### PR TITLE
feat: move precompile verification to SVM

### PR DESCRIPTION
**Pull request by jstarry**
#### Problem
Transactions with precompile signatures that fail verification are currently not allowed to be committed to a block. After https://github.com/solana-foundation/solana-improvement-documents/pull/159 they will be allowed to be committed so that block producers always get paid fees when verifying precompile signatures.

#### Summary of Changes
- Introduced a new feature gate to turn off precompile verification where we currently doing it in the banking stage and replay stage and move it to the SVM instead
- Expanded the scope of instruction processing metrics to include precompiles

Fixes #

### CODE REVIEW

File: sdk/src/precompiles.rs [hunk 1]
```diff
@@ -3,14 +3,16 @@
...
     lazy_static::lazy_static,
+    num_derive::{FromPrimitive, ToPrimitive},
     solana_decode_error::DecodeError,
+    solana_program::instruction::CompiledInstruction,
     thiserror::Error,
 };
 
 /// Precompile errors
-#[derive(Error, Debug, Clone, PartialEq, Eq)]
+#[derive(Error, Debug, Clone, PartialEq, Eq, FromPrimitive, ToPrimitive)]
```
Comment (apfitzge) [inline sdk/src/precompiles.rs:15]: I think I'm missing where we need these new traits

**Review by apfitzge — DISMISSED** [review]
Change makes sense to me. Users should still be able to verify that precompiles are valid through the interfaces, though they may need to change how they are doing it. THat should be communicated before feature-gate activation.

File: program-runtime/src/invoke_context.rs [hunk 2]
```diff
@@ -465,6 +466,38 @@ impl<'a> InvokeContext<'a> {
...
+        self.push()?;
+
+        let feature_set = self.get_feature_set();
+        let move_precompile_verification_to_svm =
+            feature_set.is_active(&feature_set::move_precompile_verification_to_svm::id());
+        if move_precompile_verification_to_svm {
+            let instruction_datas: Vec<_> = message_instruction_datas_iter.collect();
+            precompile
+                .verify(precompile_instruction_data, &instruction_datas, feature_set)
+                .map_err(InstructionError::from)
```
Comment (jstarry) [inline program-runtime/src/invoke_context.rs:490]: The new primitive traits are needed to convert to/from `InstructionError::Custom` and the `ToPrimitive` trait is specifically used here.

See this implementation for details:
```rust
impl<T> From<T> for InstructionError
where
    T: ToPrimitive,
```

**Comment by jstarry** [comment]
@Lichtso could you please check the message processor and invoke context changes?

File: sdk/src/precompiles.rs (hunk 1 above)
Comment (jstarry) [reply sdk/src/precompiles.rs:15]: https://github.com/anza-xyz/agave/pull/2441#discussion_r1708280395

File: svm/src/message_processor.rs [hunk 3]
```diff
@@ -87,53 +83,48 @@ impl MessageProcessor {
...
-                        .total_us,
-                    time
-                );
-                result
+            let mut compute_units_consumed = 0;
+            let (result, process_instruction_us) = measure_us!({
+                if let Some(precompile) = get_precompile(program_id, |feature_id| {
+                    invoke_context.get_feature_set().is_active(feature_id)
+                }) {
+                    invoke_context.process_precompile(
```
Comment (Lichtso) [inline svm/src/message_processor.rs:91]: Arguments could be reordered to be the same as `process_instruction()`

Comment (jstarry) [reply svm/src/message_processor.rs:91]: Sounds good, reordered to be more consistent here: https://github.com/anza-xyz/agave/pull/2441/commits/40cea54c99be99b5ae13fec17c12b922e4e7fcfd

**Comment by Lichtso** [comment]
Instead of using `InstructionError::Custom` can we either add a new `InstructionError::PrecompileError` or map `PrecompileError` to the existing `InstructionError`s?

**Comment by jstarry** [comment]
No I think it's more appropriate to use custom errors here because these errors aren't general, they are program specific. Can you explain why adding new instruction error variants would make more sense?

**Comment by Lichtso** [comment]
I might have missed it, but the changes in errors are not mentioned in the features SIMD, or are they?
Thus, I assumed they should stay close to the way they are now.

**Comment by jstarry** [comment]
> I might have missed it, but the changes in errors are not mentioned in the features SIMD, or are they? Thus, I assumed they should stay close to the way they are now.

No they are not but they probably should, thanks for bringing that up. There really isn't a good precedent for precompile error handling since we generally disregard any errors for dropped transactions. Currently we map precompile errors to `InstructionError::InvalidAccountIndex` inside `SanitizedTransaction::verify_precompiles` which I didn't feel was appropriate now that we will be storing precompile errors in the ledger.

I will update the SIMD with details about the errors and have them mapped to `InstructionError::Custom` as I have it here in this PR and then we can move forward with this PR as is. Sound good?

**Comment by jstarry** [comment]
Any other feedback here @Lichtso and @apfitzge?

**Comment by jstarry** [comment]
@Lichtso @apfitzge still waiting on feedback here

File: programs/ed25519-tests/tests/process_transaction.rs [hunk 4]
```diff
@@ -68,7 +70,9 @@ async fn test_failure() {
     assert_matches!(
         client.process_transaction(transaction).await,
         Err(BanksClientError::TransactionError(
-            TransactionError::InvalidAccountIndex
+            TransactionError::InstructionError(0, InstructionError::Custom(3))
         ))
     );
+    // this assert is for documenting the matched error code above
```
Comment (apfitzge) [inline programs/ed25519-tests/tests/process_transaction.rs:108]: @jstarry sorry I missed your last comment.

We should still have a test when the feature is disabled.
Is that tested somewhere else that I am missing?

**Comment by apfitzge** [comment]
think we ought to push on the SIMD as well, but may not have luck merging until after breakpoint

**Review by Lichtso — DISMISSED** [review]

File: programs/ed25519-tests/tests/process_transaction.rs (hunk 4 above)
Comment (jstarry) [reply programs/ed25519-tests/tests/process_transaction.rs:108]: Sorry no it's not, I added an extra test for that here: https://github.com/anza-xyz/agave/pull/2441/commits/1084cfab8e21c9ae40245f1982e1e2a6d574a92a

Comment (apfitzge) [reply programs/ed25519-tests/tests/process_transaction.rs:108]: perfect

**Review by apfitzge — APPROVED** [review]

**Pull request merged**